### DISCUSSION TITLE
Respource duplication in new pipelined rendering

### DISCUSSION THREAD

**Discussion by vtimonin** [bevyengine/bevy#2963] [Q&A] [answered]
I briefly checked new pipelined rendering and it seems RenderStage::Extract stage clones all heavy resources such as Images or Meshes. This looks very wasteful...

**Comment by DJMcNab** [comment, marked as answer]
We've discussed this before on [discord](https://discord.gg/bevy), and iirc, it was only adding the new items based on the events.

See e.g. 
https://discord.com/channels/691052431525675048/743663924229963868/891477293506117642

**Comment by vtimonin** [comment]
Yes, resource extraction is performed only for changed resources, that is fine. But the render world stores copies of such resources. I can be wrong and probably e.g. images are removed from CPU memory after corresponding GPU buffers are created - it should be fine in this case. 
Otherwise it can practically double used RAM.

**Reply by DJMcNab** [reply]
However, they get drained from the extracted assets here:

https://github.com/bevyengine/bevy/blob/43e8a156fb4739abee3060c53140f4a94cab9ae5/pipelined/bevy_render2/src/render_asset.rs#L143

**Comment by vtimonin** [comment]
Yes, prepare_assets consumes assets, so it seems to be ok.

**Reply by DJMcNab** [reply]
If you're satisfied (and it's possible to do so), would you be able to move this to the Q&A category and mark an answer?

Otherwise, we might need @cart to move it - he might be the only one with permission.

Also, in future please reply within a thread rather than creating a new thread for each comment.
